readme.txt
-------------------
Revision 2021-03-16
-------------------

- pseudo-tty support: edit ~/.scantoolrc and add comport_number = 1000 + pty
  number; for instance, for /dev/pts/3, set the following in ~/.scantoolrc:

  [comm]
  comport_number = 1003

Note: when setting a pseudo-tty, only use ~/.scantoolrc and do not edit the
serial port through the "Options" panel.

Compilation with Ubuntu 20.04 LTS:

sudo apt install liballegro4.4 liballegro4-dev allegro4-doc
make clean
make -e RELEASE=yes

Compilation with Ubuntu, including logging to "comm_log.txt":

make clean
make -e LOG=yes

-------------------
Original readme.txt
-------------------

*********************************************************************
   ScanTool.net OBD-II Software v1.20 for ElmScan and OBDLink devices
   Copyright (C) 2010 ScanTool.net LLC, All Rights Reserved
*********************************************************************

======================================
============= Disclaimer =============
======================================

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option) any later
version. This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
more details.

======================================
============ Introduction ============
======================================

ScanTool.net OBD-II Software for ElmScan is free software that allows you to 
use your computer and an inexpensive hardware interface to read information 
from your car's computer. Current version allows you to read trouble codes and 
see their descriptions, clear the codes and turn off the "Check Engine" light, 
and display real-time sensor data such as RPM, Engine Load, Vehicle Speed, 
Coolant Temperature, and Timing Advance.

For more information about the ElmScan OBD-II interface, please visit 
our website, located at http://www.ScanTool.net.

======================================
===== Minimum System Requirements ====
======================================

Windows:
   - 486DX 25Mhz Processor or better
   - 4Mb RAM
   - Windows 95 or higher
   - DirectX 7 or newer
   - 640x480 display
   - Serial port

DOS:
   - 386SX 10Mhz Processor or better
   - 1Mb RAM
   - DOS v3.0
   - 640x480 display
   - Serial port

======================================
========== Troubleshooting ===========
======================================

If you are having problems running the software, please visit our
Support Page located at http://www.scantool.net/support

======================================
========== Version History ===========
======================================

   v1.21  -  Made 115.2k baud rate the default
   v1.20  -  Added dynamic COM port selector (now works with COM port numbers > COM8)
             Fixed program freeze when opening Bluetooth COM ports
             Fixed potential data errors due to ELM327 UART silicone errata
             Implemented firmware version printing for new FW releases of OBDLink and OBDLink CI
             A few small bugfixes
   v1.15  -  Added support for ElmScan 5 Compact
             Added support for OBDLink and OBDLink CI devices
             Defaults can now be set by simply deleting the CFG file
             Fixed sensor pages with all sensors disabled causing other pages to delay first refresh by up to 10 seconds
             Fixed 100% CPU usage
   v1.14  -  Added 115.2k & 230.4k baud rates to options (May 23, 2008)
             Added genuine scan tool validation
   v1.13  -  Fixed some formatting, precision, and conversion errors in sensor formulas
             Fixed inconsistent refresh rate display in Sensor Data
             Refined handling of number of DTCs reported being different from number of DTCs read
             Small bug fixes
   v1.12  -  Refined error descriptions for ELM327
             CAN DTC responses are now handled properly
             SAE-defined DTCs with no description are now not identified as manufacturer-defined
             Expanded Clear Codes warning to describe all things that will be cleared
             Misc. bug fixes
   v1.11  -  Fixed a bug in Display Codes dialog, where the program would crash under some conditions
             Refined "COM Port Could not be Open" error handling
   v1.10  -  Added baud rate switching
             Made compatible with ELM327
             Added protocol detection to reset interface dialog
             Added pending DTCs
             DTCs are now read even if ECU reports 0 codes
             Added OBD Information dialog
             Cleaned up Main Menu
             Updated About dialog
             Misc. bug/cosmetic fixes
   v1.09  -  Fixed erroneous interpretation of 7F responses (KWP2000)
             Fixed number of codes reporting in Trouble Codes
             Fixed possible incorrect DTC interpretation when more than one response is received
   v1.08  -  Fixed problem with ECUs that pad the response with 0's
   v1.07  -  Added the rest of the sensors defined in SAE J1979 (APR2002)
             Added the rest of "designed to comply with" to Sensor Data
             Cleaned up portions of the code
             Added COM ports 5-8 to Options
             Updated codes.dat with latest generic P and U codes, and removed B and C codes
             System information dialog supports a wider range of processors and platforms
   v1.06  -  Fixed some problems with RS232
             Corrected air flow rate formula (US system)
             Added 16 new sensors
             Modified the layout of "Sensor Data"
             Misc. bug fixes
   v1.04  -  Updated serial library
             Some bug fixes and enhancements
             Support for multiple platforms and compilers
   v1.03  -  Fixed incorrect display of some sensors when ELM323 is used
   v1.02  -  Minor bug fixes
   v1.01  -  Minor bug fixes
   v1.00  -  Initial release


debian/changelog
scantool (2.2-0) stable; urgency=medium

  * Ported to Ubuntu 20.04 LTS
  * Pseudo-tty support
  * Bug fixing
  * Updated readme.txt

 -- Ircama Sun 20 Mar 2021, 01:12:02, CET

scantool (2.1-2) unstable; urgency=medium

  * debian/watch: Actually fix URL this time.

 -- Kees Cook <kees@debian.org>  Sun, 15 Mar 2020 20:22:26 -0700

scantool (2.1-1) unstable; urgency=medium

  * New upstream version:
    - debian/patches/libm-link.patch: incorporated upstream
  * debian/watch: Fix missing URL elements.

 -- Kees Cook <kees@debian.org>  Sat, 14 Mar 2020 09:16:30 -0700

scantool (2.0-2) unstable; urgency=medium

  * debian/patches/libm-link.patch:
    - add missing linkage against libm (Closes: #935527), thanks to
      Andreas Ronnquist <gusnan@debian.org>, and Steve Langasek
      <steve.langasek@ubuntu.com>.
  * debian/control:
    - bumps standards version to 4.5.0: no changes

 -- Kees Cook <kees@debian.org>  Fri, 06 Mar 2020 19:41:27 -0800

scantool (2.0-1) unstable; urgency=medium

  * Update to latest "upstream" release, which incorporates all
    debian patches. Most notable change is the new serial port
    code (Closes: #919201), and improved trouble code text.
  * debian/control:
    - bump standards version: compat to 12.
    - update homepage
  * debian/rules:
    - clean up redundant comments
    - remove dzcomm entirely (Closes: 920787)
  * debian/watch: aim at github

 -- Kees Cook <kees@debian.org>  Sun, 10 Mar 2019 15:31:25 -0700

scantool (1.21+dfsg-7) unstable; urgency=low

  * Fix FTCBFS: Let dh_auto_configure pass --host to ./configure,
    thanks to Helmut Grohne. (Closes: #915275)

 -- Kees Cook <kees@debian.org>  Sun, 13 Jan 2019 09:30:59 -0800

scantool (1.21+dfsg-6) unstable; urgency=medium

  * debian/control: Build-depend on liballegro4-dev, thanks to Adam Conrad.
  * debian/rules: handle already-cleaned tree more gracefully.

 -- Kees Cook <kees@debian.org>  Mon, 24 Aug 2015 16:29:13 -0700

scantool (1.21+dfsg-5) unstable; urgency=low

  * Solved FTBFS with gcc-5 (Closes: 778110):
    - explicitly use static inlines:
      - debian/patches/dzcomm-usb-serial.patch
    - fix asprintf declaration and usage:
      - debian/patches/multi-code.patch
      - debian/patches/scantool-linux-build.patch
  * Fixed lintian issues:
    - debian/copyright: add missing license short name.
    - debian/control: bump standards version, no changes needed.

 -- Kees Cook <kees@debian.org>  Sat, 14 Mar 2015 10:53:58 -0700

scantool (1.21+dfsg-4) unstable; urgency=low

  * debian/patches/scantool-linux-build.patch: update to include path to
    codes file, which should not be static.
  * debian/codes: add vendor-specific DTC code lists to available codes.
  * debian/patches/multi-code.patch: handle duplicate codes and report
    code origin as a prefix.

 -- Kees Cook <kees@debian.org>  Sun, 23 Dec 2012 10:15:51 -0800

scantool (1.21+dfsg-3) unstable; urgency=low

  * Add debian/patches/kfreebsd-build-fixes.patch: fix up source for
    kfreebsd builds.
  * debian/rules, debian/patches/build-flags.patch: pass compiler flags to
    build correctly.
  * debian/control, debian/compat: bump compat level to 9.
  * debian/control: bump standards version; no changes needed.
  * debian/patches/drop-report-request.patch: do not report to scantool.net
    since they don't want reports on this ancient software any more. :)
  * debian/patches/allow-elm327-clones.patch: fix clones harder
    (Closes: 674179).

 -- Kees Cook <kees@debian.org>  Fri, 01 Jun 2012 11:07:05 -0700

scantool (1.21+dfsg-2) unstable; urgency=low

  * Add debian/patches/allow-elm327-clones.patch: do not abort when faced
    with an ELM327 device that wasn't manufactured by ScanTool.Net
    (Closes: 630603).

 -- Kees Cook <kees@debian.org>  Wed, 22 Jun 2011 22:27:26 -0700

scantool (1.21+dfsg-1) unstable; urgency=low

  * Initial release (Closes: 627635).
    - dropped unused non-free "listports.{c,h}" from orig tarball.

 -- Kees Cook <kees@debian.org>  Sat, 28 May 2011 06:57:37 -0700


serial.c
#include <string.h>
#include <ctype.h>
#include "globals.h"
#include <allegro/internal/aintern.h>
#ifdef ALLEGRO_WINDOWS
   #include <winalleg.h>
#elif TERMIOS
   #include <stdio.h>
   #include <termios.h>
#else
   #include <dzcomm.h>
#endif
#include "serial.h"

#ifdef ALLEGRO_WINDOWS
   static HANDLE com_port;
#elif TERMIOS
   static int fdtty;
   static struct termios oldtio,newtio;
#else
   static comm_port *com_port;
#endif

#define TX_TIMEOUT_MULTIPLIER    0
#define TX_TIMEOUT_CONSTANT      1000

volatile int serial_time_out;
volatile int serial_timer_running;
struct COMPORT comport;

//timer interrupt handler for sensor data
static void serial_time_out_handler()
{
   serial_time_out = TRUE;
   serial_timer_running = FALSE;
}
END_OF_STATIC_FUNCTION(serial_time_out_handler)


void start_serial_timer(int delay)
{
   stop_serial_timer();
   install_int_ex(serial_time_out_handler, MSEC_TO_TIMER(delay));  // install the timer
   serial_time_out = FALSE;
   serial_timer_running = TRUE;
}


void stop_serial_timer()
{
   remove_int(serial_time_out_handler);
   serial_time_out = FALSE;
   serial_timer_running = FALSE;
}


void serial_module_init()
{
#ifndef ALLEGRO_WINDOWS
#ifndef TERMIOS
   dzcomm_init();
#endif
#endif
   serial_timer_running = FALSE;
   /* all variables and code used inside interrupt handlers must be locked */
   LOCK_VARIABLE(serial_time_out);
   LOCK_FUNCTION(serial_time_out_handler);
   _add_exit_func(serial_module_shutdown, "serial_module_shutdown");
}


void serial_module_shutdown()
{
   close_comport();
   
#ifndef ALLEGRO_WINDOWS
#ifndef TERMIOS
    dzcomm_closedown();
#endif
#endif
   
   _remove_exit_func(serial_module_shutdown);
}


int open_comport()
{
#ifdef ALLEGRO_WINDOWS
   DCB dcb;
   COMMTIMEOUTS timeouts;
   DWORD bytes_written;
   char temp_str[16];
#endif
   
   if (comport.status == READY)    // if the comport is open,
      close_comport();    // close it
   
#ifdef ALLEGRO_WINDOWS
   // Naming of serial ports 10 and higher: See
   // http://www.connecttech.com/KnowledgeDatabase/kdb227.htm
   // http://support.microsoft.com/?id=115831
   sprintf(temp_str, "\\\\.\\COM%i", comport.number + 1);
   com_port = CreateFile(temp_str, GENERIC_READ | GENERIC_WRITE, 0, 0, OPEN_EXISTING, 0, 0);
   if (com_port == INVALID_HANDLE_VALUE)
   {
      comport.status = NOT_OPEN; //port was not open
      return -1; // return error
   }
   
   // Setup comport
   GetCommState(com_port, &dcb);
   dcb.BaudRate = comport.baud_rate;
   dcb.ByteSize = 8;
   dcb.StopBits = ONESTOPBIT;
   dcb.fParity = FALSE;
   dcb.Parity = NOPARITY;
   dcb.fOutxCtsFlow = FALSE;
   dcb.fOutxDsrFlow = FALSE;
   dcb.fOutX = FALSE;
   dcb.fInX = FALSE;
   dcb.fDtrControl = DTR_CONTROL_ENABLE;
   dcb.fRtsControl = RTS_CONTROL_ENABLE;
   dcb.fDsrSensitivity = FALSE;
   dcb.fErrorChar = FALSE;
   dcb.fAbortOnError = FALSE;
   SetCommState(com_port, &dcb);
   
   // Setup comm timeouts
   timeouts.ReadIntervalTimeout = MAXWORD;
   timeouts.ReadTotalTimeoutMultiplier = 0;
   timeouts.ReadTotalTimeoutConstant = 0;
   timeouts.WriteTotalTimeoutMultiplier = TX_TIMEOUT_MULTIPLIER;
   timeouts.WriteTotalTimeoutConstant = TX_TIMEOUT_CONSTANT;
   SetCommTimeouts(com_port, &timeouts);
   // Hack to get around Windows 2000 multiplying timeout values by 15
   GetCommTimeouts(com_port, &timeouts);
   if (TX_TIMEOUT_MULTIPLIER > 0)
      timeouts.WriteTotalTimeoutMultiplier = TX_TIMEOUT_MULTIPLIER * TX_TIMEOUT_MULTIPLIER / timeouts.WriteTotalTimeoutMultiplier;
   if (TX_TIMEOUT_CONSTANT > 0)
      timeouts.WriteTotalTimeoutConstant = TX_TIMEOUT_CONSTANT * TX_TIMEOUT_CONSTANT / timeouts.WriteTotalTimeoutConstant;
   SetCommTimeouts(com_port, &timeouts);
   
   // If the port is Bluetooth, make sure device is active
   PurgeComm(com_port, PURGE_TXCLEAR|PURGE_RXCLEAR);
   WriteFile(com_port, "?\r", 2, &bytes_written, 0);
   if (bytes_written != 2)  // If Tx timeout occured
   {
      PurgeComm(com_port, PURGE_TXCLEAR|PURGE_RXCLEAR);
      CloseHandle(com_port);
      comport.status = NOT_OPEN; //port was not open
      return -1;
   }
#elif TERMIOS
    char tmp[54];
    if( comport.number < 100 )
        snprintf(tmp, sizeof(tmp), "/dev/ttyS%d", comport.number);
    else
        snprintf(tmp, sizeof(tmp), "/dev/ttyUSB%d", comport.number-100);
    fdtty = open( tmp, O_RDWR | O_NOCTTY );
    if (fdtty <0) { return(-1); }

    tcgetattr(fdtty,&oldtio); /* save current port settings */

    bzero(&newtio, sizeof(newtio));

    cfsetspeed(&newtio, comport.baud_rate);

    cfmakeraw(&newtio);
    newtio.c_cflag |= (CLOCAL | CREAD);

            // No parity (8N1):
    newtio.c_cflag &= ~PARENB;
    newtio.c_cflag &= ~CSTOPB;
    newtio.c_cflag &= ~CSIZE;
    newtio.c_cflag |= CS8;

        // disable hardware flow control
    newtio.c_cflag &= ~CRTSCTS ;

    newtio.c_cc[VTIME]    = 0;   /* inter-character timer unused */
    newtio.c_cc[VMIN]     = 0;   /* blocking read until 5 chars received */

    tcflush(fdtty, TCIFLUSH);
    tcsetattr(fdtty,TCSANOW,&newtio);
#else
   com_port = comm_port_init(comport.number);
   if (!com_port) {
      write_log(szDZCommErr);
      comport.status = NOT_OPEN;
      return -1;
   }
   comm_port_set_baud_rate(com_port, comport.baud_rate);
   comm_port_set_parity(com_port, NO_PARITY);
   comm_port_set_data_bits(com_port, BITS_8);
   comm_port_set_stop_bits(com_port, STOP_1);
   comm_port_set_flow_control(com_port, NO_CONTROL);
   if (comm_port_install_handler(com_port) != 1)
   {
      write_log(szDZCommErr);
      comport.status = NOT_OPEN; //port was not open
      return -1; // return error
   }
#endif
   
   serial_time_out = FALSE;
   comport.status = READY;
   
   return 0; // everything is okay
}


void close_comport()
{
   if (comport.status == READY)    // if the comport is open, close it
   {
#ifdef ALLEGRO_WINDOWS
      PurgeComm(com_port, PURGE_TXCLEAR|PURGE_RXCLEAR);
      CloseHandle(com_port);
#elif TERMIOS
      tcsetattr(fdtty,TCSANOW,&oldtio);
      close(fdtty);
#else
      comm_port_flush_output(com_port);
      comm_port_flush_input(com_port);
      comm_port_uninstall(com_port);
#endif
   }
   comport.status = NOT_OPEN;
}


void send_command(const char *command)
{
   char tx_buf[32];
#ifdef ALLEGRO_WINDOWS
   DWORD bytes_written;
#endif
   
   sprintf(tx_buf, "%s\r", command);  // Append CR to the command
   
#ifdef ALLEGRO_WINDOWS
   PurgeComm(com_port, PURGE_TXCLEAR|PURGE_RXCLEAR);
   WriteFile(com_port, tx_buf, strlen(tx_buf), &bytes_written, 0);
   if (bytes_written != strlen(tx_buf))
   {
#ifdef LOG_COMMS
      log_comm("TX ERROR", tx_buf);  // Log transmission error
#endif
      return;
   }
#elif TERMIOS
//    printf("tx:'%s'\n",command);
	if( write(fdtty,tx_buf,strlen(tx_buf)) == -1 )
	{
	  perror("write tty");
	  close(fdtty);
	  fdtty = -1;
	}
	else
	{
        tcflush(fdtty, TCIFLUSH);
	}
#else
   comm_port_flush_output(com_port);
   comm_port_flush_input(com_port);
   comm_port_string_send(com_port, tx_buf);
#endif
   
#ifdef LOG_COMMS
   write_comm_log("TX", tx_buf);
#endif
}


int read_comport(char *response)
{
   char *prompt_pos = NULL;
#ifdef ALLEGRO_WINDOWS
   DWORD bytes_read = 0;
   DWORD errors;
   COMSTAT stat;
   int i, j;
   
   response[0] = '\0';
   ClearCommError(com_port, &errors, &stat);
   if (stat.cbInQue > 0)
      ReadFile(com_port, response, stat.cbInQue, &bytes_read, 0);
   
   // Remove extraneous 0s
   for (i = 0, j = 0; i < bytes_read; i++)
      if (response[i] > 0)
         response[j++] = response[i];
   response[j] = 0;
#elif TERMIOS
   int res=1;
   fd_set readfs;
   struct timeval timeout;
   char tmp[64];
   bzero(response,64);
   bzero(tmp,64);

   FD_ZERO(&readfs);
   FD_SET(fdtty, &readfs);
   while( res != 0 && fdtty != -1)
   {
       timeout.tv_usec = 200e3;  /* millisecondes */
       timeout.tv_sec  = 0;  /* secondes */
       res = select(fdtty+1, &readfs, NULL, NULL, &timeout);
       if( res != 0 )
       {
           if( read(fdtty, tmp, 64) != 0 )
           {
               strcat( response, tmp );
               bzero(tmp,64);
           }
       }
   }

//    printf("rx:'");
//    i=0;while(response[i]!=0) { printf("%c",response[i]>=32?response[i]:'.');i++;}
//    printf("'\n");

#else
   int i;
   char c;
   
   i = 0;
   while((c = comm_port_test(com_port)) >= 0) // while the serial buffer is not empty, read comport
      if (c > 0)
         response[i++] = c;
   response[i] = 0;
#endif
   
   prompt_pos = strchr(response, '>');
   if (prompt_pos != NULL)
   {
#ifdef LOG_COMMS
      write_comm_log("RX", response);
#endif
      *prompt_pos = '\0'; // erase ">"
      return PROMPT;      // command prompt detected
   }
   else if (strlen(response) == 0)  // if the string is empty,
      return EMPTY;
   else                         //otherwise,
   {
#ifdef LOG_COMMS
      write_comm_log("RX", response);
#endif
      return DATA;
   }
}


int find_valid_response(char *buf, char *response, const char *filter, char **stop)
{
   char *in_ptr = response;
   char *out_ptr = buf;

   buf[0] = 0;

   while (*in_ptr)
   {
      if (strncmp(in_ptr, filter, strlen(filter)) == 0)
      {
         while (*in_ptr && *in_ptr != SPECIAL_DELIMITER) // copy valid response into buf
         {
            *out_ptr = *in_ptr;
            in_ptr++;
            out_ptr++;
         }
         *out_ptr = 0;  // terminate string
         if (*in_ptr == SPECIAL_DELIMITER)
            in_ptr++;
         break;
      }
      else
      {
         // skip to the next delimiter
         while (*in_ptr && *in_ptr != SPECIAL_DELIMITER)
            in_ptr++;
         if (*in_ptr == SPECIAL_DELIMITER)  // skip the delimiter
            in_ptr++;
      }
   }

   if (stop)
      *stop = in_ptr;

   if (strlen(buf) > 0)
      return TRUE;
   else
      return FALSE;
}

// DO NOT TRANSLATE ANY STRINGS IN THIS FUNCTION!
int process_response(const char *cmd_sent, char *msg_received)
{
   int i = 0;
   char *msg = msg_received;
   int echo_on = TRUE; //echo status
   int is_hex_num = TRUE;
   char temp_buf[80];

   if (cmd_sent)
   {
      for(i = 0; cmd_sent[i]; i++)
      {
         if (cmd_sent[i] != *msg)    // if the characters are not the same,
         {
            echo_on = FALSE;  // say that echo is off
            break;            // break out of the loop
         }
         msg++;
      }

      if (echo_on == TRUE)  //if echo is on
      {
         send_command("ate0"); // turn off the echo
         start_serial_timer(AT_TIMEOUT);
         // wait for chip response or timeout
         while ((read_comport(temp_buf) != PROMPT) && !serial_time_out)
            ;
         stop_serial_timer();
         if (!serial_time_out)
         {
            send_command("atl0"); // turn off linefeeds
            start_serial_timer(AT_TIMEOUT);
            // wait for chip response or timeout
            while ((read_comport(temp_buf) != PROMPT) && !serial_time_out)
               ;
            stop_serial_timer();
         }
      }
      else //if echo is off
         msg = msg_received;
   }

   while(*msg && (*msg <= ' '))
      msg++;

   if (strncmp(msg, "SEARCHING...", 12) == 0)
      msg += 13;
   else if (strncmp(msg, "BUS INIT: OK", 12) == 0)
      msg += 13;
   else if (strncmp(msg, "BUS INIT: ...OK", 15) == 0)
      msg += 16;

   for(i = 0; *msg; msg++) //loop to copy data
   {
      if (*msg > ' ')  // if the character is not a special character or space
      {
         if (*msg == '<') // Detect <DATA_ERROR
         {
            if (strncmp(msg, "<DATA ERROR", 10) == 0)
               return DATA_ERROR2;
            else
               return RUBBISH;
         }
         msg_received[i] = *msg; // rewrite response
         if (!isxdigit(*msg) && *msg != ':')
            is_hex_num = FALSE;
         i++;
      }
      else if (((*msg == '\n') || (*msg == '\r')) && (msg_received[i-1] != SPECIAL_DELIMITER)) // if the character is a CR or LF
         msg_received[i++] = SPECIAL_DELIMITER; // replace CR with SPECIAL_DELIMITER
   }
   
   if (i > 0)
      if (msg_received[i-1] == SPECIAL_DELIMITER)
         i--;
   msg_received[i] = '\0'; // terminate the string

   if (is_hex_num)
      return HEX_DATA;

   if (strcmp(msg_received, "NODATA") == 0)
      return ERR_NO_DATA;
   if (strcmp(msg_received + strlen(msg_received) - 15, "UNABLETOCONNECT") == 0)
      return UNABLE_TO_CONNECT;
   if (strcmp(msg_received + strlen(msg_received) - 7, "BUSBUSY") == 0)
      return BUS_BUSY;
   if (strcmp(msg_received + strlen(msg_received) - 9, "DATAERROR") == 0)
      return DATA_ERROR;
   if (strcmp(msg_received + strlen(msg_received) - 8, "BUSERROR") == 0 ||
       strcmp(msg_received + strlen(msg_received) - 7, "FBERROR") == 0)
      return BUS_ERROR;
   if (strcmp(msg_received + strlen(msg_received) - 8, "CANERROR") == 0)
      return CAN_ERROR;
   if (strcmp(msg_received + strlen(msg_received) - 10, "BUFFERFULL") == 0)
      return BUFFER_FULL;
   if (strncmp(msg_received, "BUSINIT:", 8) == 0)
   {
      if (strcmp(msg_received + strlen(msg_received) - 5, "ERROR") == 0)
         return BUS_INIT_ERROR;
      else
         return SERIAL_ERROR;
   }
   if (strcmp(msg_received, "?") == 0)
      return UNKNOWN_CMD;
   if (strncmp(msg_received, "ELM320", 6) == 0)
      return INTERFACE_ELM320;
   if (strncmp(msg_received, "ELM322", 6) == 0)
      return INTERFACE_ELM322;
   if (strncmp(msg_received, "ELM323", 6) == 0)
      return INTERFACE_ELM323;
   if (strncmp(msg_received, "ELM327", 6) == 0)
      return INTERFACE_ELM327;
   if (strncmp(msg_received, "OBDLink", 7) == 0 ||
       strncmp(msg_received, "STN1000", 7) == 0 ||
       strncmp(msg_received, "STN11", 5) == 0)
      return INTERFACE_OBDLINK;
   if (strncmp(msg_received, "SCANTOOL.NET", 12) == 0)
      return STN_MFR_STRING;
   if (strcmp(msg_received, "OBDIItoRS232Interpreter") == 0)
      return ELM_MFR_STRING;
   
   return RUBBISH;
}


int display_error_message(int error, int retry)
{
   char buf[32];
   
   switch (error)
   {
      case BUS_ERROR:
         return alert("Bus Error: OBDII bus is shorted to Vbatt or Ground.", NULL, NULL, (retry) ? "Retry" : "OK", (retry) ? "Cancel" : NULL, 0, 0);

      case BUS_BUSY:
         return alert("OBD Bus Busy. Try again.", NULL, NULL, (retry) ? "Retry" : "OK", (retry) ? "Cancel" : NULL, 0, 0);

      case BUS_INIT_ERROR:
         return alert("OBD Bus Init Error. Check connection to the vehicle,", "make sure the vehicle is OBD-II compliant,", "and ignition is ON.", (retry) ? "Retry" : "OK", (retry) ? "Cancel" : NULL, 0, 0);

      case UNABLE_TO_CONNECT:
         return alert("Unable to connect to OBD bus.", "Check connection to the vehicle. Make sure", "the vehicle is OBD-II compliant, and ignition is ON.", (retry) ? "Retry" : "OK", (retry) ? "Cancel" : NULL, 0, 0);

      case CAN_ERROR:
         return alert("CAN Error.", "Check connection to the vehicle. Make sure", "the vehicle is OBD-II compliant, and ignition is ON.", (retry) ? "Retry" : "OK", (retry) ? "Cancel" : NULL, 0, 0);

      case DATA_ERROR:
      case DATA_ERROR2:
         return alert("Data Error: there has been a loss of data.", "You may have a bad connection to the vehicle,", "check the cable and try again.", (retry) ? "Retry" : "OK", (retry) ? "Cancel" : NULL, 0, 0);

      case BUFFER_FULL:
         return alert("Hardware data buffer overflow.", NULL, NULL, (retry) ? "Retry" : "OK", (retry) ? "Cancel" : NULL, 0, 0);

      case SERIAL_ERROR:
      case UNKNOWN_CMD:
      case RUBBISH:
         return alert("Serial Link Error: please check connection", "between computer and scan tool.", NULL, (retry) ? "Retry" : "OK", (retry) ? "Cancel" : NULL, 0, 0);

      default:
         sprintf(buf, "Unknown error occured: %i", error);
         return alert(buf, NULL, NULL, (retry) ? "Retry" : "OK", (retry) ? "Cancel" : NULL, 0, 0);
   }
}


const char *get_protocol_string(int interface_type, int protocol_id)
{
   switch (interface_type)
   {
      case INTERFACE_ELM320:
         return "SAE J1850 PWM (41.6 kBit/s)";
      case INTERFACE_ELM322:
         return "SAE J1850 VPW (10.4 kBit/s)";
      case INTERFACE_ELM323:
         return "ISO 9141-2 / ISO 14230-4 (KWP2000)";
      case INTERFACE_ELM327:
      case INTERFACE_OBDLINK:
         switch (protocol_id)
         {
            case 0:
               return "N/A";
            case 1:
               return "SAE J1850 PWM (41.6 kBit/s)";
            case 2:
               return "SAE J1850 VPW (10.4 kBit/s)";
            case 3:
               return "ISO 9141-2";
            case 4:
               return "ISO 14230-4 KWP2000 (5-baud init)";
            case 5:
               return "ISO 14230-4 KWP2000 (fast init)";
            case 6:
               return "ISO 15765-4 CAN (11-bit ID, 500 kBit/s)";
            case 7:
               return "ISO 15765-4 CAN (29-bit ID, 500 kBit/s)";
            case 8:
               return "ISO 15765-4 CAN (11-bit ID, 250 kBit/s)";
            case 9:
               return "ISO 15765-4 CAN (29-bit ID, 250 kBit/s)";
         }
   }
   
   return "unknown";
}
